### DISCUSSION TITLE
"High level doc" for integrating into a new editor?

### DISCUSSION THREAD

**Discussion by mauricioszabo** [tree-sitter/tree-sitter#2061] [Q&A]
Hi, hope I'm not spamming too much this discussion.

Basically, I am having trouble following up the documentation on how I'm supposed to actually _integrate_ tree-sitter into an editor. I am using `tree-sitter-web`.

From what I understand following up the [syntax highlight documentation](https://tree-sitter.github.io/tree-sitter/syntax-highlighting#overview) on the official site, seems that I need to _query_ what I have parsed and with that I'll have "highlight names" like `function.buitin`. Is this correct?

But then, on [Local Variables](https://tree-sitter.github.io/tree-sitter/syntax-highlighting#local-variables), seems that queries are different: I have a `local.reference` that _may refer_ to an earlier definition - how exactly do I use this?

Using the Ruby code on the same page as example, for me it returns the first `list` as BOTH a `local.definition` and a `local.reference`, and the second `list` as just a `local.reference`. But it also says that `process_list` and `map`, etc are `local.reference`s too. So, how should that be used?

**Comment by ahelwer** [comment]
Don't know whether `tree-sitter highlight` functionality is built in to the bindings you're using; reference highlighting is tricky and nvim-treesitter rolls their own implementation. The built-in reference highlighting isn't great, there's no way to bump a definition up to the parent scope for example.

**Reply by mauricioszabo** [reply]
But what actually is being implemented in nvim-treesitter? What exactly are the steps that need to be done?

I'm asking because so far I saw implementations that only used one query to highlight things, others that used 3 to 4, others that used no query at all, others that parsed a JSON, etc.....

I'm not a rust developer, so it's also tricky for me to read the code to highlight and understand the steps that I need to implement myself (my bindings does not expose highlight either it seems)

**Reply by ahelwer** [reply]
Completely dependent on the editor and what you're trying to do with the grammar. What are you trying to do?

**Reply by mauricioszabo** [reply]
Ok, so:

1. I have a source code in an editor
2. I want to highlight that code, the same as tree-sitter highlight do
3. I already know how to colorize things on this specific editor, so there's nothing editor specific here

I just want to know WHAT I have to do. Currently I am parsing the source, then querying highlights.scm, and using this result as the "tokens" to reflect on the editor.

Problem - this is different from what tree-sitter highlight shows. I saw some documentation on locals.scm, but I honestly don't know how to use it...

**Reply by ahelwer** [reply]
You'll have to implement locals yourself. Usually this involves declaring some nodes are scopes and some nodes are definitions and some nodes are references, then identifying which reference refers to which definition and highlighting it appropriately.

**Reply by mauricioszabo** [reply]
I was actually querying the locals.scm file, but the results are confusing and I have no idea on how to actually use them...

**Reply by ahelwer** [reply]
Queries are just queries. The names mean nothing, they aren't special. They're just a label saying something at some location in the file is called whatever you chose. It's up to you to do something with them. Here's an example algorithm:

1. Create a hierarchical tree of all scopes
2. For each definition, find the first parent scope. Associate that definition with that scope and all child scopes of that scope.
3. For each reference, find the first parent scope. Check whether a matching definition exists in that scope. If so, highlight the reference the same way the definition is highlighted.

You'll quickly notice that associating definitions with the first parent scope isn't sufficient, and you'll often want to bump the definition up to the scope above that, or even the global scope. Neovim uses predicates for this.

**Reply by mauricioszabo** [reply]
Yes, I understand that queries are just queries - my question is how to handle when a query returns multiple matches / captures for the same range. 

As an example: here's how `tree-sitter highlight` works on this specific case:
![image](https://user-images.githubusercontent.com/138037/215286706-c9c02e58-d0da-433c-a805-10ef7151aadb.png)

Looking at my config, it _seems that_ `process_list` is a `function`, `context` is a variable, and that both `list` are `variable.parameter` (because if I delete these specific keys from my tree-sitter config, they stop being highlighted on my console). I also tried to add `variable` as a parameter, and yes, `context` started to be highlighted as a variable (just for sanity check).

Now - I am trying to replicate this highlight on my editor. I queried using the `highlights.scm` file, the same source, and used `.captures` from `tree-sitter-web` packages. Then I checked for `names` and the node's text, and here's what I got:

```js
locals.map(e => [e.name, e.node.text, e.node.startIndex, e.node.endIndex])
// => 
0: (4) ["keyword", "def", 0, 3]
1: (4) ["function.method", "process_list", 4, 16]
2: (4) ["function.method", "process_list", 4, 16]
3: (4) ["variable", "process_list", 4, 16]
4: (4) ["punctuation.bracket", "(", 16, 17]
5: (4) ["variable.parameter", "list", 17, 21]
6: (4) ["function.method", "list", 17, 21]
7: (4) ["variable", "list", 17, 21]
8: (4) ["punctuation.bracket", ")", 21, 22]
9: (4) ["function.method", "context", 25, 32]
10: (4) ["variable", "context", 25, 32]
11: (4) ["operator", "=", 33, 34]
12: (4) ["function.method", "current_context", 35, 50]
13: (4) ["variable", "current_context", 35, 50]
14: (4) ["function.method", "list", 53, 57]
15: (4) ["variable", "list", 53, 57]
16: (4) ["punctuation.delimiter", ".", 57, 58]
17: (4) ["function.method", "map", 58, 61]
18: (4) ["function.method", "map", 58, 61]
19: (4) ["variable", "map", 58, 61]
20: (4) ["punctuation.bracket", "{", 62, 63]
21: (4) ["punctuation.bracket", "}", 63, 64]
22: (4) ["keyword", "end", 65, 68]
```

Interesting points are: `process_list` is highlighted as `function.method` twice, and as `variable`
But then, `context` is also highlighted as `function.method` (once) as as `variable`. So why is `process_list` highlighted as function, and `context` as `variable`?

**Reply by mauricioszabo** [reply]
Ok, sorry - I _may know_ what's happening. It seems that `tree-sitter-web` have a problem with `refutedProperties`. I was using a version someone posted with a fix, but it basically made all `refutedProperties` disappear - when I parse with `node-tree-sitter` the query result is different: patterns on index `2` and `9` have `refutedProperties: {local: null}`, so if I do ignore the index `9` then, indeed, `context` is only highlighted as `variable`. Am I right on my interpretation?

**Reply by ahelwer** [reply]
If the queries aren't working the way you want then you can rewrite them. Most integrations write their own set of queries for each grammar. Neovim queries are different from highlight command queries for my grammar. Some entities will match multiple queries and it is up to you to develop a system for which label takes precedence.

**Reply by mauricioszabo** [reply]
@ahelwer sure, but again, before I write my own queries I wanted to at least make the highlighting work correctly with the queries I have now... the queries that I am using are the same as the ones tree-sitter highlight uses, so for now I would prefer to understand how these things are supposed to work, then make changes if I need them...

**Reply by ahelwer** [reply]
I don't even think anybody or anything uses tree-sitter highlight queries. As far as I can tell they're useful as a reference for what other integrations could define for their own highlight queries.

**Reply by ahlinc** [reply]
> I don't even think anybody or anything uses tree-sitter highlight queries. As far as I can tell they're useful as a reference for what other integrations could define for their own highlight queries.

At least they are used by `tree-sitter highlight` command :slightly_smiling_face:
Also maybe Github highlighting somehow use them but I'm not sure, @dcreager and @maxbrunsfeld may be know something more about this.

**Reply by mauricioszabo** [reply]
Yes, `tree-sitter highlight` command that comes with tree-sitter uses the default queries that come with tree-sitter.

With that out of the way, how can I make the _exact same highlight_ that the `tree-sitter highlight` offers? Again, I'm not asking for a way to make colors on my editor, or even the API/commands that I need to use on the WASM version - just the steps, like "query this file, use only the first/the last/specific token to highlight, then query this second file...."

**Reply by ahelwer** [reply]
I already gave you the algorithm up above. Reference highlighting overrides other types of highlighting. If multiple non reference queries match the same token then the one that comes first in the file is used. Neovim it's the opposite, later queries override earlier ones.

**Reply by mauricioszabo** [reply]
Yes, but this is not what the `tree-sitter highlight` seems to be _actually doing_ for me - when I query for things, there are two problematic tokens that should be highlighted the same but are not:

```ruby
def process_list(list)
  context = current_context
  list.map {}
end
```

On `tree-sitter highlight` command-line: `process_list` is highlighted as `function`, `context` is highlighted as `variable`, and `current_context` as `function`.

But when I query for these, that's not what I get back in return. Supposing I queried with `captures = query.captures(parsed.rootNode)`, here's the result I get:

```js
captures.map(e => [e.node.text, e.name, (e.refutedProperties && e.refutedProperties.local === null)])
0: (3) ["def", "keyword", undefined]
1: (3) ["process_list", "function.method", undefined]
2: (3) ["process_list", "function.method", true]
3: (3) ["process_list", "variable", undefined]
4: (3) ["(", "punctuation.bracket", undefined]
5: (3) ["list", "variable.parameter", undefined]
6: (3) ["list", "function.method", true]
7: (3) ["list", "variable", undefined]
8: (3) [")", "punctuation.bracket", undefined]
9: (3) ["context", "function.method", true]
10: (3) ["context", "variable", undefined]
11: (3) ["=", "operator", undefined]
12: (3) ["current_context", "function.method", true]
13: (3) ["current_context", "variable", undefined]
14: (3) ["list", "function.method", true]
15: (3) ["list", "variable", undefined]
16: (3) [".", "punctuation.delimiter", undefined]
17: (3) ["map", "function.method", undefined]
18: (3) ["map", "function.method", true]
19: (3) ["map", "variable", undefined]
20: (3) ["{", "punctuation.bracket", undefined]
21: (3) ["}", "punctuation.bracket", undefined]
22: (3) ["end", "keyword", undefined]
```

So, indexes 9 and 10 say that `context` if `function.method` and then, a `variable`. `tree-sitter highlight` command-line says it's a `variable`, so it's choosing the _second_ item. Then, on indexes 12 and 13 we have the results for `current_context`, and they are _exactly the same_ as `context`, except that now `tree-sitter highlight` interprets this as `variable`. So what's exactly going on here?

**Reply by mauricioszabo** [reply]
Also, by "Reference highlighting" are you referring to the "locals.scm" query?

**Reply by ahlinc** [reply]
> Also, by "Reference highlighting" are you referring to the "locals.scm" query?

There was description of different `scm` files: https://github.com/tree-sitter/tree-sitter/discussions/1251#discussioncomment-987520

**Reply by ahlinc** [reply]
> So, indexes 9 and 10 say that `context` if `function.method` and then, a `variable`.

On your output there is no ranges and without them it's hard to realize how different captures relate to each other and what of them may overlap. Please use a `tree-sitter query` for example snippets.
<details>

![Screenshot from 2023-01-29 20-08-31](https://user-images.githubusercontent.com/14666676/215346705-899a46f0-ac8a-4a06-a574-74d3a6226aad.png)

</details>

**Reply by mauricioszabo** [reply]
Ok, here it is (thought that it was clear I was already filtering only the valid ranges, but ok):

<details>
<summary>cat /tmp/code.rb</summary>

```
def process_list(list)
  context = current_context
  list.map do |item|
    process_item(item, context)
  end
end
```
</details>

<details>
<summary>tree-sitter q --scope source.ruby tree-sitter-ruby/queries/highlights.scm /tmp/code.rb</summary>

```
/tmp/algo.rb
  pattern: 0
    capture: 0 - keyword, start: (0, 0), end: (0, 3), text: `def`
  pattern: 7
    capture: 2 - function.method, start: (0, 4), end: (0, 16), text: `process_list`
  pattern: 24
    capture: 2 - function.method, start: (0, 4), end: (0, 16), text: `process_list`
  pattern: 25
    capture: 9 - variable, start: (0, 4), end: (0, 16), text: `process_list`
  pattern: 36
    capture: 20 - punctuation.bracket, start: (0, 16), end: (0, 17), text: `(`
  pattern: 20
    capture: 8 - variable.parameter, start: (0, 17), end: (0, 21), text: `list`
  pattern: 24
    capture: 2 - function.method, start: (0, 17), end: (0, 21), text: `list`
  pattern: 25
    capture: 9 - variable, start: (0, 17), end: (0, 21), text: `list`
  pattern: 36
    capture: 20 - punctuation.bracket, start: (0, 21), end: (0, 22), text: `)`
  pattern: 24
    capture: 2 - function.method, start: (1, 2), end: (1, 9), text: `context`
  pattern: 25
    capture: 9 - variable, start: (1, 2), end: (1, 9), text: `context`
  pattern: 34
    capture: 18 - operator, start: (1, 10), end: (1, 11), text: `=`
  pattern: 24
    capture: 2 - function.method, start: (1, 12), end: (1, 27), text: `current_context`
  pattern: 25
    capture: 9 - variable, start: (1, 12), end: (1, 27), text: `current_context`
  pattern: 24
    capture: 2 - function.method, start: (2, 2), end: (2, 6), text: `list`
  pattern: 25
    capture: 9 - variable, start: (2, 2), end: (2, 6), text: `list`
  pattern: 35
    capture: 19 - punctuation.delimiter, start: (2, 6), end: (2, 7), text: `.`
  pattern: 4
    capture: 2 - function.method, start: (2, 7), end: (2, 10), text: `map`
  pattern: 24
    capture: 2 - function.method, start: (2, 7), end: (2, 10), text: `map`
  pattern: 25
    capture: 9 - variable, start: (2, 7), end: (2, 10), text: `map`
  pattern: 0
    capture: 0 - keyword, start: (2, 11), end: (2, 13), text: `do`
  pattern: 16
    capture: 8 - variable.parameter, start: (2, 15), end: (2, 19), text: `item`
  pattern: 24
    capture: 2 - function.method, start: (2, 15), end: (2, 19), text: `item`
  pattern: 25
    capture: 9 - variable, start: (2, 15), end: (2, 19), text: `item`
  pattern: 4
    capture: 2 - function.method, start: (3, 4), end: (3, 16), text: `process_item`
  pattern: 24
    capture: 2 - function.method, start: (3, 4), end: (3, 16), text: `process_item`
  pattern: 25
    capture: 9 - variable, start: (3, 4), end: (3, 16), text: `process_item`
  pattern: 36
    capture: 20 - punctuation.bracket, start: (3, 16), end: (3, 17), text: `(`
  pattern: 24
    capture: 2 - function.method, start: (3, 17), end: (3, 21), text: `item`
  pattern: 25
    capture: 9 - variable, start: (3, 17), end: (3, 21), text: `item`
  pattern: 35
    capture: 19 - punctuation.delimiter, start: (3, 21), end: (3, 22), text: `,`
  pattern: 24
    capture: 2 - function.method, start: (3, 23), end: (3, 30), text: `context`
  pattern: 25
    capture: 9 - variable, start: (3, 23), end: (3, 30), text: `context`
  pattern: 36
    capture: 20 - punctuation.bracket, start: (3, 30), end: (3, 31), text: `)`
  pattern: 0
    capture: 0 - keyword, start: (4, 2), end: (4, 5), text: `end`
  pattern: 0
    capture: 0 - keyword, start: (5, 0), end: (5, 3), text: `end`
```

</details>

<details>
<summary>tree-sitter q --scope source.ruby tree-sitter-ruby/queries/locals.scm /tmp/code.rb </summary>

```
/tmp/code.rb
  pattern: 0
    capture: local.scope, start: (0, 0), end: (5, 3)
  pattern: 11
    capture: 2 - local.reference, start: (0, 4), end: (0, 16), text: `process_list`
  pattern: 7
    capture: 1 - local.definition, start: (0, 17), end: (0, 21), text: `list`
  pattern: 11
    capture: 2 - local.reference, start: (0, 17), end: (0, 21), text: `list`
  pattern: 12
    capture: 1 - local.definition, start: (1, 2), end: (1, 9), text: `context`
  pattern: 11
    capture: 2 - local.reference, start: (1, 2), end: (1, 9), text: `context`
  pattern: 11
    capture: 2 - local.reference, start: (1, 12), end: (1, 27), text: `current_context`
  pattern: 11
    capture: 2 - local.reference, start: (2, 2), end: (2, 6), text: `list`
  pattern: 11
    capture: 2 - local.reference, start: (2, 7), end: (2, 10), text: `map`
  pattern: 1
    capture: local.scope, start: (2, 11), end: (4, 5)
  pattern: 3
    capture: 1 - local.definition, start: (2, 15), end: (2, 19), text: `item`
  pattern: 11
    capture: 2 - local.reference, start: (2, 15), end: (2, 19), text: `item`
  pattern: 11
    capture: 2 - local.reference, start: (3, 4), end: (3, 16), text: `process_item`
  pattern: 11
    capture: 2 - local.reference, start: (3, 17), end: (3, 21), text: `item`
  pattern: 11
    capture: 2 - local.reference, start: (3, 23), end: (3, 30), text: `context`
```

</details>

So, my results are the same as yours, it seems, but with a different format. So these are the scopes from the highlights query, that I don't know how to filter which one will need to be used to highlight:

```
  pattern: 24
    capture: 2 - function.method, start: (1, 2), end: (1, 9), text: `context`
  pattern: 25
    capture: 9 - variable, start: (1, 2), end: (1, 9), text: `context`
  pattern: 34
    capture: 18 - operator, start: (1, 10), end: (1, 11), text: `=`
  pattern: 24
    capture: 2 - function.method, start: (1, 12), end: (1, 27), text: `current_context`
  pattern: 25
    capture: 9 - variable, start: (1, 12), end: (1, 27), text: `current_context`
```

And the ones for locals:

```
  pattern: 12
    capture: 1 - local.definition, start: (1, 2), end: (1, 9), text: `context`
  pattern: 11
    capture: 2 - local.reference, start: (1, 2), end: (1, 9), text: `context`
  pattern: 11
    capture: 2 - local.reference, start: (1, 12), end: (1, 27), text: `current_context`
```

So, do I basically ignore anything that comes from "highlights" and check if there's a `local.definition` on locals, and basically "force" it to be `variable`? That's what I don't actually get...

**Reply by ahlinc** [reply]
It's up to developer how to interpret a query result.
By rearranging `matching patterns` in a query file it's possible to make an impact on what `pattern` matches would appear first if there are multiple patterns that match _the same range_ in an input.
Usually developer may think about a `sliding window` that moves forward by some following match or stay on a same range _(in this case usually the first match is the most specific and following matches that cover the same range may be skipped)_. There even may be a situation when a following match may cover _shorter range_ that a current match, but this usually more looks like a bug in query patterns definition.

**Reply by ahlinc** [reply]
> So, do I basically ignore anything that comes from "highlights" and check if there's a local.definition on locals, and basically "force" it to be variable? That's what I don't actually get...

The `locals.scm` file is not intended to use for highlighting directly. Patterns in the locals file match a huge multi line spans of code for ex. a function body or a whole class definition. Matches in the `locals.scm` file may help to mange a state in a code that some other highlighting matches are in a context let say in a context of a function body and in that case you may want to color them differntly.

Note: the `tree-sitter highlight` concatenates the files and executes them like a one query:
https://github.com/tree-sitter/tree-sitter/blob/6f42c78f2d0173f64d9346b30f0e8c46b5a9b7a6/highlight/src/lib.rs#L188-L194

<details><summary>But keep in mind the order matters</summary>

![Screenshot from 2023-01-29 21-18-11](https://user-images.githubusercontent.com/14666676/215350513-1dc0b445-da08-40e7-91ad-c9b8516f064c.png)
</details>
